File: BootCampOOP/Car/Car.cpp
#include "Point.h"
#include "Car.h"

Car::Car(double capacityIn, double consumptionIn, const Point& locationIn, const std::string& modelIn) {

    fuelCapacity = capacityIn;
    fuelConsumption = consumptionIn;
    location = locationIn;
    model = modelIn;
    fuelAmount = 30;
    
}

Car::~Car(){
    
}


double Car::getFuelAmount() const{
    return fuelAmount;
   
}

double Car::getFuelCapacity() const{
    return fuelCapacity;
    
}

double Car::getFuelConsumption() const{
    return fuelConsumption;
    
}

const Point& Car::getLocation() const{
    return location;
    
}

const std::string& Car::getModel() const{
    return model;
    
}

void Car::drive(const Point& destination){
    
    double distance = destination.distance(location);
    
    double requiredFuel = distance * fuelConsumption;
    if (fuelAmount < requiredFuel) {
        throw OutOfFuel();
    }
    else {
        location = destination;
        fuelAmount -= requiredFuel;
        
    }
    
    
}
    
void Car::drive(double x, double y){
    
    drive(Point(x,y));
    
}

void Car::refill(double fuel){
    
    if (fuelAmount + fuel > fuelCapacity) {
        throw ToMuchFuel();
    }
    else {
        fuelAmount += fuel;
    }
    
    
}

std::ostream& operator<<(std::ostream& out, const Car& car){
    out << "" << std::endl;
    out << "Модель:" << car.getModel() << std::endl;
    out << "Обьём бака:" << car.getFuelCapacity() << std::endl;
    out << "Расход топлива:" << car.getFuelConsumption() << std::endl;
    out << "Кол-во топлива:" << car.getFuelAmount() << std::endl;
    out << "Местоположение" << car.getLocation() << std::endl;
    
    return out;
}

File: BootCamp Java/Car/Car.java
public class Car {
    private double fuelAmount;
    private double fuelCapacity;
    private double fuelConsumption;
    private Point location;
    private String model;
    
    public Car(double capacity, double consumption, Point location, String model) {
        this.fuelCapacity = capacity;
        this.fuelConsumption = consumption;
        this.model = model;
        this.fuelAmount = 0;
        this.location = location;
        this.model = model;
    }

    public Car() {
        this.fuelCapacity = 60;
        this.fuelConsumption = 0.6;
        this.location = new Point();
        this.model = "Mercedes";
    }
    public double getFuelAmount () {
        return this.fuelAmount;
    }

    public double getFuelCapacity () {
        return this.fuelCapacity;
    }

    public double getFuelConsumption () {
        return this.fuelConsumption;
    }

    public Point getLocation() {
        return this.location;
    }

    public String getModel() {
        return this.model;
    }

    public void drive( Point destination) throws OutOfFuelException {
        double temp_consumption = location.distance(destination) * this.fuelConsumption;
        if ( this.fuelAmount < temp_consumption ) {
            throw new OutOfFuelException();
        }
        this.fuelAmount -= temp_consumption;
        this.location = destination;
    }

    public void drive(double x, double y) throws OutOfFuelException {
        Point destination = new Point (x, y);
        
        drive(destination);
    }

    public void refill(double fuel) throws ToMuchFuelException{
        if ( this.fuelAmount + fuel > fuelCapacity ) {
            throw new ToMuchFuelException();
        }
        this.fuelAmount += fuel;
    }

    public String toString() {
        return "Car model:"+getModel()+'\n'+"Capacity:"+getFuelCapacity()+'\n'+"Fuel consumption:"+getFuelConsumption()+'\n'+"Current fuel amount:"+getFuelAmount()+'\n'+"Current fuel consumption:"+getFuelConsumption();
    }
}
